packages/deployments/contracts/contracts/SponsorVault.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.11;
+
+import {ISponsorVault} from "./interfaces/ISponsorVault.sol";
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20, Address} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+interface ITokenExchange {
+  function getInGivenExactOut(address token, uint256 exactOut) external returns (uint256);
+  function swapExactOut(address token, uint256 exactOut, address recipient) external payable returns (uint256);
+}
+
+interface IGasTokenOracle {
+  function getRate(uint32 originDomain) external view returns (uint256 num, uint256 den);
+}
+
+/**
+ * @title SponsorVault
+ * @author Connext Labs
+ * @notice Contains logic for sponsoring liquidity and relayer fees
+ */
+contract SponsorVault is ISponsorVault, Ownable{
+  // ============ Libraries ============
+  using SafeERC20 for IERC20;
+
+  // ============ Struct ============
+  struct Rate {
+    uint256 num;
+    uint256 den;
+  }
+
+  // ============ Private storage ============
+
+  // ============ Public storage ============
+
+  /**
+   * @notice The address of connext
+   */
+  address public connext;
+
+  /**
+   * @notice The origin domain to this domain gas token rates
+   * @dev Used when no oracle is available
+   */
+  mapping(uint32 => Rate) public rates;
+
+  /**
+   * @notice The maximum amount of relayer fee to sponsor
+   */
+  uint256 public relayerFeeCap;
+
+  /**
+   * @notice The origin domain to this domain gas token oracle
+   * @dev Used to calculate sponsored relayer fee
+   */
+  address public gasTokenOracle;
+
+  /**
+   * @notice The this domain gas token to token exchange
+   * @dev Used to calculate sponsored relayer fee
```

Comment:
need some more clarification in this comment as well -- is this a dex you can use to get a price or is it a dex used to swap out some tokens? im not sure why you would need to swap anything, you should just need to fund this vault with lots of native asset, and then some way to get the `origin domain native asset : current domain native asset` price

Reply:
ahhhh okay, so it swaps into the right asset? should have some global slippage set by the owner in that case as well then

Reply:
Fixed the comment, it was wrong.
The tokenExchange would be used to get the token used to pay the liquidity fee swapping in the current domain native asset wouldn't the `maxAmountIn` be implicitly defined in the `value` sent when calling `swapExactOut` ?

Reply:
> The tokenExchange would be used to get the token used to pay the liquidity fee using the current domain native asset to swap in, so wouldn't the maxAmountIn be implicitly defined in the value sent when calling swapExactOut ?

im not sure i follow what you're saying here 😅 . Let me restate to test:

The `tokenExchange` is used to get the token that the liquidity fee is charged in, using the native asset stored on this contract. And you're saying that the `maxAmountIn` is implicitly defined as the `value` when you're sending to the `swapExactOut` call?

I guess that should work, if I'm understanding correctly. How would you know that the `value` is enough for the `swapExactOut` (I'm using the assumption that the `swapExactOut` means you get the exact value out of the AMM, so the value you put *in* could vary)?

Reply:
The main critique is just that the contract should have "sensible bounds". Though I could definitely see the argument where you're giving this money away anyways so maybe its okay to have really generous slippage tolerances

Reply:
>The tokenExchange is used to get the token that the liquidity fee is charged in, using the native asset stored on this contract. And you're saying that the maxAmountIn is implicitly defined as the value when you're sending to the swapExactOut call?

yes

> I guess that should work, if I'm understanding correctly. How would you know that the value is enough for the swapExactOut (I'm using the assumption that the swapExactOut means you get the exact value out of the AMM, so the value you put in could vary)?

The `ITokenExchange` has a `getInGivenExactOut` function, to get the amount you need to send. 

Now that I think it better it may not work, bc if the amount being sent to the AMM is not enough due to slippage the swap would probably revert due to the exact out, so now I think is would be better to use a `swapExactIn` and use the resulting amount as the sponsor.

Reply:
Agree!